pennylane/transforms/optimization/commute_controlled.py

```diff
@@ -21,6 +27,277 @@
 from .optimization_utils import find_next_gate
 
 
+@lru_cache
+def _get_plxpr_commute_controlled():  # pylint: disable=missing-function-docstring,too-many-statements
+    try:
+        # pylint: disable=import-outside-toplevel
+        from jax import make_jaxpr
+
+        from pennylane.capture import PlxprInterpreter
+        from pennylane.capture.primitives import measure_prim
+    except ImportError:  # pragma: no cover
+        return None, None
+
+    # pylint: disable=redefined-outer-name
+
+    class CommuteControlledInterpreter(PlxprInterpreter):
+        """Plxpr Interpreter for applying the ``commute_controlled`` transform to callables or jaxpr
+        when program capture is enabled.
+
+        .. note::
+            If the direction is set to ``"right"``, this class interprets the operations by scanning them
+            backward after the jaxpr has been traversed (pushing the gates to the right of controlled gates).
+            This is because we can only traverse the jaxpr in a forward direction with the current implementation when
+            program capture is enabled.
+
+            This is less efficient than setting the direction to ``"left"``, which allows the interpreter
+            to push the gates to the left of controlled gates as it interprets the jaxpr.
+
+            Despite this, the default direction is ``"right"`` to maintain compatibility with the
+            current default value of the transform.
+        """
+
+        def __init__(self, direction: Optional[str] = "right"):
+            """Initialize the interpreter."""
+
+            if direction not in ("left", "right"):
+                raise ValueError("Direction for commute_controlled must be 'left' or 'right'.")
+
+            self.direction = direction
+            self.op_deque = deque()
+            self._env = {}
+            self.current_index = 0
+
+        def setup(self) -> None:
+            """Initialize the instance before interpreting equations."""
+            self.op_deque.clear()
+
+        def cleanup(self) -> None:
+            """Clean up the instance after interpreting equations."""
+            self.op_deque.clear()
+
+        def _interpret_operation_left(self, op: Operator) -> list:
+            """Interpret a PennyLane operation and push it through controlled operations as far left as possible."""
+
+            # This function follows the same logic used in the `_commute_controlled_left` function.
+
+            if not _can_be_pushed_through(op):
+                self.current_index += 1
+                self.op_deque.append(op)
+                return []
+
+            prev_gate_idx = _find_previous_gate_on_wires(op.wires, self.op_deque)
+            new_index = self.current_index
+
+            while prev_gate_idx is not None:
+                prev_gate = self.op_deque[new_index - prev_gate_idx - 1]
```

Comment:
Could also do this to make it clear how you are adjusting the next index,
```suggestion
                prev_gate = self.op_deque[new_index - (prev_gate_idx + 1)]
```